tasks/Task.cpp

```diff
@@ -2,48 +2,496 @@
 
+Task::~Task() {}
 
 /// The following lines are template definitions for the various state machine
 // hooks defined by Orocos::RTT. See Task.hpp for more detailed
 // documentation about them.
 
 bool Task::configureHook()
 {
-    if (! TaskBase::configureHook())
+    if (!TaskBase::configureHook())
         return false;
+
+    mConfig.iceServers.emplace_back(_stun_server.get());
+    mWs = std::make_shared<rtc::WebSocket>();
+
+    configureWebSocket();
+
     return true;
 }
 bool Task::startHook()
 {
-    if (! TaskBase::startHook())
+    if (!TaskBase::startHook())
         return false;
+
+    if (!_remote_peer_id.get().empty())
+    {
+        std::future<void> wait_remote_peer_future = mWaitRemotePeerPromise.get_future();
+        mRemotePeerID = _remote_peer_id.get();
+        mRemotePeerAnswerReceived = false;
+
+        base::Time start_time = base::Time::now();
+        // Get contact with the remote peer and create datachannel
+        while (!mRemotePeerAnswerReceived)
+        {
+            base::Time duration = base::Time::now() - start_time;
+            if (duration > _wait_remote_peer_time_out.get())
+            {
+                mWaitRemotePeerPromise.set_exception(
+                    std::make_exception_ptr(std::runtime_error("Timeout to find the remote peer")));
+                break;
+            }
+            ping();
+            sleep_for(100ms);
+        }
+        wait_remote_peer_future.get();
+
+        std::future<void> dc_future = mDataChannelPromise.get_future();
+        mPeerConnection = initiatePeerConnection();
+        string label = _data_channel_label.get();
+        mDataChannel = mPeerConnection->createDataChannel(label);
+        registerDataChannelCallBacks(mDataChannel);
+        dc_future.get();
+    }
+    else
+    {
+        std::future<void> dc_future = mDataChannelPromise.get_future();
+        base::Time start_time = base::Time::now();
+        // Wait for the datachannel open until the timeout
+        while (mState.data_channel != DataChannelOpened)
+        {
+            base::Time duration = base::Time::now() - start_time;
+            if (duration > _data_channel_open_time_out.get())
+            {
+                mDataChannelPromise.set_exception(
+                    std::make_exception_ptr(std::runtime_error("Timeout to open datachannel")));
+                break;
+            }
+        }
+        dc_future.get();
+    }
+
     return true;
 }
 void Task::updateHook()
 {
+    _status.write(mState);
+
+    evaluateDataChannel();
+    evaluateWebSocket();
+
     TaskBase::updateHook();
 }
-void Task::errorHook()
-{
-    TaskBase::errorHook();
-}
+void Task::errorHook() { TaskBase::errorHook(); }
 void Task::stopHook()
 {
+    if (mDataChannel->isOpen())
+    {
+        std::future<void> dc_close_future = mDataChannelClosePromise.get_future();
+        mDataChannel->close();
+        dc_close_future.get();
+    }
+    if (mPeerConnection)
```

Comment:
I would have expected you to follow the same pattern in both ifs ... either open/close or delete.

Reply:
There is not isOpen method for the peerConnection...I will just check the pointer in both cases.